### ISSUE TITLE
Integrate TUK into the SOTI STM32 Project

### ISSUE THREAD

**Issue by Koloss0** [UMSATS/soti#13]
This will be necessary for SOTI to communicate with other subsystems with the new changes.

Todo:
 - [ ] Delete the old CAN code.
 - [ ] Initialise CAN Wrapper with `CANWrapper_Init` (See the [Wiki](https://github.com/UMSATS/tsat-utilities-kit/wiki/CAN-Wrapper-Module-Usage-Guide)). Make sure ACKs are **not** ignored.
 - [ ] Create a message handler function which simply serializes incoming messages (both ACKs and non-ACKs) and sends them to the CLI over UART.
 - [ ] Make sure you are polling messages and errors in the main loop. Create a stud error handler for now.
 - [ ] Update the UART interrupt code to use `CANWrapper_Transmit` to forward messages instead.

**Comment by ArnavG-it** [comment]
SOTI has its own [CAN queue implementation](https://github.com/UMSATS/soti/blob/main/soti-stm32project-boardV1/Core/Inc/can_message_queue.h), which looks identical to TUK's aside from being able to be dynamically sized on initialisation. Should this be replaced with the queue from TUK as well?

**Comment by Koloss0** [comment]
> SOTI has its own [CAN queue implementation](https://github.com/UMSATS/soti/blob/main/soti-stm32project-boardV1/Core/Inc/can_message_queue.h), which looks identical to TUK's aside from being able to be dynamically sized on initialisation. Should this be replaced with the queue from TUK as well?

Yes, those are remnants of the old API, before it became the CAN Wrapper Module. The [can.c](https://github.com/UMSATS/soti/blob/main/soti-stm32project-boardV1/Core/Src/can.c) file is also part of that. Also, SOTI's queue implementation is [not the same](https://github.com/UMSATS/soti/blob/main/soti-stm32project-boardV1/Core/Src/can_message_queue.c) as the TUK one, as it embeds commands for the LCD display (among other things, like how it uses [malloc](https://github.com/UMSATS/soti/blob/main/soti-stm32project-boardV1/Core/Src/can_message_queue.c#L29) when it shouldn't). I consider that bad practice (the queue implementation shouldn't be concerned with the LCD display), so I don't have a problem if that code gets thrown out, too. If we want to re-implement the LCD code, that should be done outside the queue implementation.

In conclusion, yes the old code should be replaced with TUK's implementation. I've already imported the submodule, so you basically just need to call the init function in main.

**Comment by ArnavG-it** [comment]
347eb65 adds the init code (mostly temporary), preceded by 4dc5328 adding the include path. 

Some questions:
1. Currently, `.node_id` in `CANWrapper_InitTypeDef` must be defined and either be NODE_CDH, POWER, ADCS, or PAYLOAD in order to initialise successfully. What's the solution to incorporate SOTI? (An `#ifdef` to add NODE_SOTI at compilation maybe?)
2. I get the error "_conflicting types for 'CANWrapper_Init'; have 'int()'_" when calling [CANWrapper_Init](https://github.com/UMSATS/soti/commit/347eb655e20e80e83468abba90541d018f5de943#diff-561214c24b23c24eed43ab6ad149423456af9f53958ce6e58ecf067fc7a0b145R88), and a warning in `can_wrapper.h` of the same function: "_previous declaration of 'CANWrapper_Init' with type 'CANWrapper_StatusTypeDef(CANWrapper_InitTypeDef)'_". Not sure what's up. Any ideas?

Thanks!

**Comment by Koloss0** [comment]
> [347eb65](https://github.com/UMSATS/soti/commit/347eb655e20e80e83468abba90541d018f5de943) adds the init code (mostly temporary), preceded by [4dc5328](https://github.com/UMSATS/soti/commit/4dc5328b2aa5daec2aa85c3bc5abbddae1c241bf) adding the include path.
> 
> Some questions:
> 
> 1. Currently, `.node_id` in `CANWrapper_InitTypeDef` must be defined and either be NODE_CDH, POWER, ADCS, or PAYLOAD in order to initialise successfully. What's the solution to incorporate SOTI? (An `#ifdef` to add NODE_SOTI at compilation maybe?)
> 2. I get the error "_conflicting types for 'CANWrapper_Init'; have 'int()'_" when calling [CANWrapper_Init](https://github.com/UMSATS/soti/commit/347eb655e20e80e83468abba90541d018f5de943#diff-561214c24b23c24eed43ab6ad149423456af9f53958ce6e58ecf067fc7a0b145R88), and a warning in `can_wrapper.h` of the same function: "_previous declaration of 'CANWrapper_Init' with type 'CANWrapper_StatusTypeDef(CANWrapper_InitTypeDef)'_". Not sure what's up. Any ideas?
> 
> Thanks!

1. That's a good thing to bring up. We're limited to only four node IDs, so making a node ID for SOTI isn't possible. That's okay though, because that wouldn't be useful anyways. What we really want is to be able to _masquerade_ as any subsystem we want when sending commands. This is unfortunately counter to the current design of TUK, which expects you to initialise it with a node ID in advance. To fix this, we'll probably have to introduce a `CANWrapper_Set_Node_ID` function so that we can update it on the fly. It complicates the API a little bit, so we should probably put a little note saying it's for advanced usage only.

2. C doesn't support function calls outside of main. The error you're getting suggests the compiler thinks you are trying to redefine CANWrapper_Init. The call to the init function should go in the main function sometime after `MX_CAN1_Init`.

**Comment by ArnavG-it** [comment]
> What we really want is to be able to masquerade as any subsystem we want when sending commands

Interesting, that makes sense.

> C doesn't support function calls outside of main.

Haha that's so obvious now that you pointed it out. Thanks.

**Comment by ArnavG-it** [comment]
SOTI reports the metadata of received messages (see [here](https://github.com/UMSATS/soti/blob/main/soti-stm32project-boardV1/Core/Src/main.c#L432-L435) and [here](https://github.com/UMSATS/soti/blob/main/soti-stm32project-boardV1/Core/Src/can.c#L98-L106)), but only the sender and message itself are accessible through TUK's [message callback](https://github.com/UMSATS/tsat-utilities-kit/blob/main/Src/can_wrapper/can_wrapper.c#L110). Is there a way to access the remaining metadata found in the `CachedCANMessage` struct through TUK?

**Comment by Koloss0** [comment]
No there isn't. The message callback was designed with the intention of keeping things simple for the subsystems. SOTI is a special case where we want to offload all the details to the command line, whereas in the subsystem code we don't particularly care what the priority or recipient is.

We could use another `#ifdef` but that quickly leads to [preprocessor hell](https://medium.com/@deissenboeck/living-in-the-ifdef-hell-afd629e7e943). I think the simplest solution to this is to provide more metadata to the callback function.

So I imagine an API like this:

```c
void on_message_received(CANMessage msg)
{
    NodeID sender = msg.sender;
    NodeID recipient = msg.recipient;
    bool is_ack = msg.is_ack;
    uint8_t priority = msg.priority;
    CmdID cmd = msg.cmd;
    uint8_t *body = msg.body;
}
```

This would basically remove the need for the `CachedCANMessage` type as well.

**Comment by ArnavG-it** [comment]
Thanks for updating the message callback, that makes it pretty simple on SOTI's end.

Most of the work is done here: 46995ac. There's one feature that was lost:

SOTI is using dynamic queue sizes so that `satelliteToGroundQueue` holds 13000 items, and  `groundToSatelliteQueue` holds 1000. I've updated these queues to use the CANWrapper's queue implementation which is hardcoded to a size of 100. I'm not sure if this will be a problem or what the best solution would be.

Also, is [`CanQueueItem`](https://github.com/UMSATS/tsat-utilities-kit/blob/e6e9c9386bd9c61bc6a9d05f250f76c9c4067730/Inc/tuk/can_wrapper/can_queue.h#L18-L21) still needed? If so, that's fine. It just leads to a bit more work like this:
```c
void on_message_received(CANMessage msg)
{
	CANQueueItem queued_msg;
	queued_msg.msg = msg;
	CANQueue_Enqueue(&satelliteToGroundQueue, queued_msg);
}
```

**Comment by ArnavG-it** [comment]
Now that I'm looking at it, I don't think `satelliteToGroundQueue` is even needed as the message can be sent to the command line right when it's received in the callback. I removed it here: f1013fc, but it's easy to revert if we decide we need it.

`groundToSatelliteQueue` is still required as far as I can tell.

**Comment by Koloss0** [comment]
> Now that I'm looking at it, I don't think `satelliteToGroundQueue` is even needed

Correct. That's what the `message_queue` is for.

> `groundToSatelliteQueue` is still required as far as I can tell.

Correct, for now. More and more, I'm starting to believe we will need some safer way of transmitting CAN messages from an ISR. I imagine some kind of `CANWrapper_Transmit_Deferred` function and a `CANWrapper_Process_Deferred_Transmissions`.

**Comment by Koloss0** [comment]
> SOTI is using dynamic queue sizes so that `satelliteToGroundQueue` holds 13000 items, and `groundToSatelliteQueue` holds 1000. I've updated these queues to use the CANWrapper's queue implementation which is hardcoded to a size of 100. I'm not sure if this will be a problem or what the best solution would be.

We don't need dynamic allocation in this case, and in general [it's discouraged in embedded programming](https://www.reddit.com/r/embedded/comments/etrbg0/why_are_you_not_supposed_to_use_malloccalloc_in/). I'm not sure why the queues were so big before, but there's no need for it. I doubt the queues will even come close to capacity by the time they get processed in the main loop.

> Also, is [`CanQueueItem`](https://github.com/UMSATS/tsat-utilities-kit/blob/e6e9c9386bd9c61bc6a9d05f250f76c9c4067730/Inc/tuk/can_wrapper/can_queue.h#L18-L21) still needed?

It's not technically needed in that it doesn't provide any useful function. But it does decouple the queue implementation from the contents of the items. Granted, it is a CAN queue, so the item it contains probably will always just be a `CANMessage`. But, having it like this does make it very easy to add more fields if we feel like it.

Here's a trick you can use to reduce the clutter:
> 
> ```c
> void on_message_received(CANMessage msg)
> {
> 	CANQueue_Enqueue(&satelliteToGroundQueue, (CANQueueItem){ .msg = msg });
> }
> ```

**Comment by ArnavG-it** [comment]
I'll get a PR ready so the changes are all visible in one place.


One last thing: The [NodeID](https://github.com/UMSATS/tsat-utilities-kit/blob/e6e9c9386bd9c61bc6a9d05f250f76c9c4067730/Inc/tuk/can_wrapper/can_message.h#L19-L25) enum in TUK has CDH as 0 and Power as 1 whereas SOTI's sender IDs are [in reverse](https://github.com/UMSATS/soti/blob/58782df73e915946829c4c7ae2fef346b9720e06/soti-cli/cli_utils/constants.py#L78C1-L84C2). I'm assuming that these should match?

**Comment by ArnavG-it** [comment]
> What we really want is to be able to masquerade as any subsystem we want when sending commands. This is unfortunately counter to the current design of TUK, which expects you to initialise it with a node ID in advance. To fix this, we'll probably have to introduce a CANWrapper_Set_Node_ID function so that we can update it on the fly.

This should probably be a separate issue if it involves some frontend work too.

**Comment by Koloss0** [comment]
> The [NodeID](https://github.com/UMSATS/tsat-utilities-kit/blob/e6e9c9386bd9c61bc6a9d05f250f76c9c4067730/Inc/tuk/can_wrapper/can_message.h#L19-L25) enum in TUK has CDH as 0 and Power as 1 whereas SOTI's sender IDs are [in reverse](https://github.com/UMSATS/soti/blob/58782df73e915946829c4c7ae2fef346b9720e06/soti-cli/cli_utils/constants.py#L78C1-L84C2). I'm assuming that these should match?

Yes, this is another case of the code just being outdated. SOTI's enum should exactly match TUK's NodeID. I'll add an issue for this
